supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala

```diff
@@ -28,40 +33,59 @@ case class JesJobExecutionActor(override val jobDescriptor: BackendJobDescriptor
                                 serviceRegistryActor: ActorRef)
   extends BackendJobExecutionActor {
 
-  override def receive: Receive = LoggingReceive {
+  private def jesReceiveBehavior: Receive = LoggingReceive {
     case AbortJobCommand =>
       executor.foreach(_ ! AbortJobCommand)
     case abortResponse: AbortedResponse =>
       context.parent ! abortResponse
       context.stop(self)
-
-    // PBE TODO: use PartialFunction.orElse instead of this catch-all case, because Akka might use isDefinedAt over this partial function
-    case message => super.receive(message)
+    case KvPair(key, id @ Some(operationId)) if key.key == JesOperationIdKey =>
+      // Successful operation ID lookup during recover.
+      executor foreach { _ ! Recover(JesJobId(operationId))}
+    case KvKeyLookupFailed(_) =>
+      // Missed operation ID lookup during recover, fall back to execute.
+      executor foreach { _ ! Execute }
+    case KvFailure(_, e) =>
+      // Failed operation ID lookup during recover, crash and let the supervisor deal with it.
+      completionPromise.tryFailure(e)
+      throw new RuntimeException("Failure attempting to look up JES operation ID for key " + jobDescriptor.key, e)
   }
 
+  override def receive = jesReceiveBehavior orElse super.receive
+
   override val configurationDescriptor = jesConfiguration.configurationDescriptor
 
-  // PBE keep a reference to be able to hand to a successor executor if the failure is recoverable, or to complete as a
-  // failure if the failure is not recoverable.
   private lazy val completionPromise = Promise[BackendJobExecutionResponse]()
 
-  // PBE keep a reference for abort purposes.  Maybe we don't need this if we can look up actors by name.
   private var executor: Option[ActorRef] = None
 
-  // PBE there should be some consideration of supervision here.
-
-  override def recover: Future[BackendJobExecutionResponse] = ???
-
-  override def execute: Future[BackendJobExecutionResponse] = {
+  private def launchExecutor: Future[Unit] = Future {
     val executionProps = JesAsyncBackendJobExecutionActor.props(jobDescriptor,
       completionPromise,
       jesConfiguration,
       initializationData,
       serviceRegistryActor)
     val executorRef = context.actorOf(executionProps)
     executor = Option(executorRef)
-    executorRef ! Execute
-    completionPromise.future
+    ()
+  }
+
+  override def recover: Future[BackendJobExecutionResponse] = {
+    import JesJobExecutionActor._
+
+    for {
+      _ <- launchExecutor
+      _ = serviceRegistryActor ! KvGet(ScopedKey(jobDescriptor.descriptor.id, jobDescriptor.key, JesOperationIdKey))
+      c <- completionPromise.future
+    } yield c
```

Comment:
I'll point out first that @kshakir seemed to be suggesting a different approach over in his PR though I'm not sure what that approach would be.  Anyway here's how the current scheme works:

Both `recover` above and `execute` below launch an executor actor and hand back a `Future` based on the completion promise.  `execute` just tells the executor to run with an `Execute` message.  `recover` is a bit more complicated, it must first try to fetch the operation ID based on the job key before messaging the executor.  So the first thing `recover` does is query the KV store for the operation ID.

The KV store operation ID query can be answered in one of three ways: 

* `KvPair(...)` which provides an operation ID and allows for true recovery.
* `KvKeyLookupFailed(...)` where there is no operation ID found.  This is less than ideal but not a disaster, the logic just falls back to running the job anew.
* `KvFailure` where something went horribly wrong.

The `Promise` is used here as a proxy to the `Future` that is required to be returned by the `execute` and `recovery` method signatures.  Once the outcome of the recovery or execution is known, that `Promise` is completed either successfully with the response, or unsuccessfully with the reason for failure.  

In the `KvFailure` case something has gone horribly wrong, so the `Future` is completed as failed.  In the other cases the execution is proceeding normally and the existing code will complete the `Future` at a later point as it did prior to these changes.